apps/typegpu-docs/src/content/examples/rendering/ray-marching/index.ts

```diff
@@ -0,0 +1,261 @@
+      dist: sdf.sdSphere(std.sub(localP, sphere1Offset), 0.5),
+      color: d.vec3f(0.4, 1, 0.5),
+    });
+    const sphere2 = ShapeResult({
+      dist: sdf.sdSphere(std.sub(localP, sphere2Offset), 0.3),
+      color: d.vec3f(1, 0.8, 0.2),
+    });
+    const box = ShapeResult({
+      dist: sdf.sdBoxFrame(rotatedP, boxSize, 0.1),
+      color: d.vec3f(0.4, 0.6, 1.0),
+    });
+
+    // Smoothly blend shapes and colors
+    const spheres = smoothUnionColor(sphere1, sphere2, 0.1);
+    return smoothUnionColor(spheres, box, 0.2);
+  });
+
+const getSceneDist = tgpu['~unstable']
+  .fn([d.vec3f], ShapeResult)((p) => {
+    const shape = getMorphingShape(p, time.$);
+    const floor = ShapeResult({
+      dist: sdf.sdPlane(p, d.vec3f(0, 1, 0), 0),
+      color: std.mix(
+        d.vec3f(1),
+        d.vec3f(0.2),
+        checkerBoard(std.mul(p.xz, 0.1)),
+      ), // White floor
+    });
+
+    // TODO: Use regular min for floor to keep it solid white
+    return smoothUnionColor(shape, floor, 0.01);
+  });
+
+const rayMarch = tgpu['~unstable']
+  .fn([d.vec3f, d.vec3f], ShapeResult)((ro, rd) => {
+    let dO = d.f32(0);
+    const result = ShapeResult({
+      dist: d.f32(MAX_DIST),
+      color: d.vec3f(0, 0, 0),
+    });
+
+    for (let i = 0; i < MAX_STEPS; i++) {
+      const p = std.add(ro, std.mul(rd, dO));
+      const scene = getSceneDist(p);
+      dO += scene.dist;
+
+      if (dO > MAX_DIST || scene.dist < SURF_DIST) {
+        result.dist = dO;
+        result.color = scene.color;
+        break;
+      }
+    }
+
+    return result;
+  });
+
+const softShadow = tgpu['~unstable'].fn(
+  [d.vec3f, d.vec3f, d.f32, d.f32, d.f32],
+  d.f32,
+)((ro, rd, minT, maxT, k) => {
+  let res = d.f32(1);
+  let t = minT;
+
+  for (let i = 0; i < 100; i++) {
+    if (t >= maxT) break;
+    const h = getSceneDist(std.add(ro, std.mul(rd, t))).dist;
+    if (h < 0.001) return 0;
+    res = std.min(res, k * h / t);
+    t += std.max(h, 0.001);
+  }
+
+  return res;
+});
+
+const getNormal = tgpu['~unstable'].fn([d.vec3f], d.vec3f)((p) => {
+  const dist = getSceneDist(p).dist;
+  const e = 0.01;
+
+  const n = d.vec3f(
+    getSceneDist(std.add(p, d.vec3f(e, 0, 0))).dist - dist,
+    getSceneDist(std.add(p, d.vec3f(0, e, 0))).dist - dist,
+    getSceneDist(std.add(p, d.vec3f(0, 0, e))).dist - dist,
+  );
+
+  return std.normalize(n);
+});
+
+const getOrbitingLightPos = tgpu['~unstable'].fn([d.f32], d.vec3f)((t) => {
+  const radius = d.f32(3);
+  const height = d.f32(6);
+  const speed = d.f32(1);
+
+  return d.vec3f(
+    std.cos(t * speed) * radius,
+    height + std.sin(t * speed) * radius,
+    4,
+  );
+});
+
+const vertexMain = tgpu['~unstable'].vertexFn({
+  in: { idx: d.builtin.vertexIndex },
+  out: { pos: d.builtin.position, uv: d.vec2f },
+})((input) => {
+  const pos = [d.vec2f(1, 1), d.vec2f(-1, 1), d.vec2f(1, -1), d.vec2f(-1, -1)];
+  const uv = [d.vec2f(1, 1), d.vec2f(0, 1), d.vec2f(1, 0), d.vec2f(0, 0)];
```

Comment:
Doing a single triangle which covers the screen would be tiiiiiny bit more performant due to duplicated fragment work along the diagonal.

More info https://wallisc.github.io/rendering/2021/04/18/Fullscreen-Pass.html

Reply:
We're doing that in a couple examples already, I'll pull it in from those 🫡